### ISSUE TITLE
twisted.names server sends incorrect authority section

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4513]
|<img alt="loewis's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| loewis reported|
|-|-|
|Trac ID|trac#4513|
|Type|defect|
|Created|2010-06-20 21:31:23Z|
|Branch|https://github.com/twisted/twisted/tree/misbehaved-authority-4513|

In authority.py, the server sends NS records for the name in the authority section even if the client asked for a different name.

This is correct only when the NS records denote a referral. If they belong to the zone being served, including them can cause problems:
Suppose the client asks for AAAA of twistedmatrix.com. The correct answer would be empty, and the authority section should include the SOA record:

RFC 1034, 3.7, says that a SOA record may be included in the authority section; 4.3.4 says this can be used for negative caching (using the minimum value of SOA as the timeout). It says that SOA is added as an additional record, but RFC 2181, 7.1, clarifies that this is a miswording. 4.3.2 says that NS records get added to the authority section *only* if its a referral.

As a consequence, looking for an AAAA record on a zone that has none is interpreted as a referral leading to infinite recursion. The recursive resolver will eventually detect the infinite recursion and abort.

As a second (independent) issue, RFC 2181, 6.1, clarifies that NS records belong to the child zone, therefore, a referral is not authoritative.

A third issue is that referrals are not sent when the query was for ALL_RECORDS. This has no rationale, AFAICT, and RFC 1034, 4.3.2, doesn't special-case this in any way.

All three issues are fixed in attached patch.

Attachments:

* [authority.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/597/59742a8a333b3b17da581c348504652cf6d52b48/44fb31973d525079fea6e0e42acb79899fceb5d6.diff) (1590 bytes) - added by loewis on 2010-06-20 21:31:50Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4513 4513
type__defect defect
reporter__loewis loewis
priority__normal normal
milestone__ 
branch__branches_misbehaved_authority_4513 branches/misbehaved-authority-4513
branch_author__loewis__exarkun loewis, exarkun
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1277069483000000 1277069483000000
changetime__1279057546000000 1279057546000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_1"></a>|@jyknight commented|
|-|-|

Thanks! Once you (or anyone else) adds some test cases for those changes, then please add the "review" keyword so it can be reviewed and committed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [29633]) Branching to 'misbehaved-authority-4513'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun removed owner|
|-|-|

Thanks a lot for the RFC references.  They made understanding what's going on here quite a bit easier. :)

All three behaviors should be fixed in the branch now.  There are new unit tests covering the cases.  I had to make a few additional changes beyond what was in authority.diff, but not much (and overall, the fixes aren't all that complicated, certainly not compared to understanding what's supposed to be happening :).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to @exarkun|
|-|-|

Please merge!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [29820]) Merge misbehaved-authority-4513 

Author: loewis, exarkun
Reviewer: therve
Fixes: [#4513](https://github.com/twisted/twisted/issues/4513)

Stop emitting incorrect referrals from the authoritative resolver
in Twisted Names.  Also emit an SOA record when there is no matching
name to allow negative caching.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
